styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -365,17 +365,24 @@ private synchronized void tryPollPods() {
     final int resourceVersion = Integer.parseInt(list.getMetadata().getResourceVersion());
 
     for (Pod pod : list.getItems()) {
+      // Emit events
       logEvent(Watcher.Action.MODIFIED, pod, resourceVersion, true);
-      inspectPod(Watcher.Action.MODIFIED, pod);
+      final boolean isUnwantedPod = inspectPod(Watcher.Action.MODIFIED, pod);
+      final boolean isStyxPod = pod.getMetadata().getName().startsWith(STYX_RUN);
+
+      if (isUnwantedPod && isStyxPod) {
+        LOG.info("Deleting unwanted pod {}", pod.getMetadata().getName());
+        client.pods().delete(pod);
+      }
     }
   }
 
-  private void inspectPod(Watcher.Action action, Pod pod) {
+  private boolean inspectPod(Watcher.Action action, Pod pod) {
```

Comment:
Yeah, I went a bit back and forth about this and the reason I ended up having true mean the pod is unwanted is that I don't want a default value (false in the case of boolean) to indicate that a potentially dangerous action should be taken (pod deletion). It's slightly easier for a future code change to make `inspectPod` return false by mistake than it is to return true by mistake.

Still, I'm not too happy with this code, it could do with some refactoring to make things more explicit.